pkg/api/htpasswd.go

```diff
@@ -0,0 +1,203 @@
+	}
+
+	return hc.credMap.Set(login, string(passphrase))
+}
+
+// CheckPassword checks whether the user has a specified password.
+// It returns an error if the user is not found or passwords do not match,
+// and returns the nil on passwords match.
+func (hc *HtpasswdClient) CheckPassword(login, password string) error {
+	passwordHash, ok := hc.Get(login)
+	if !ok {
+		return zerr.ErrBadUser
+	}
+
+	err := bcrypt.CompareHashAndPassword([]byte(passwordHash), []byte(password))
+	if err != nil {
+		return zerr.ErrPasswordsDoNotMatch
+	}
+
+	return nil
+}
+
+// ChangePassword changes the user password.
+// It accepts user login, his supposed old password for verification and new password.
+func (hc *HtpasswdClient) ChangePassword(login, supposedOldPassword, newPassword string) error {
+	if len(newPassword) == 0 {
+		return zerr.ErrPasswordIsEmpty
+	}
+
+	oldPassphrase, ok := hc.credMap.Get(login)
+	if !ok {
+		return zerr.ErrBadUser
+	}
+
+	// given old password must match actual old password
+	if err := bcrypt.CompareHashAndPassword([]byte(oldPassphrase), []byte(supposedOldPassword)); err != nil {
+		return zerr.ErrOldPasswordIsWrong
+	}
+
+	// if passwords match, no need to update file and map, return nil as if operation is successful
+	if err := bcrypt.CompareHashAndPassword([]byte(oldPassphrase), []byte(newPassword)); err == nil {
+		return nil
+	}
+
+	// encrypt new password
+	newPassphrase, err := bcrypt.GenerateFromPassword([]byte(newPassword), bcrypt.DefaultCost)
+	if err != nil {
+		return fmt.Errorf("error occurred while encrypting new password: %w", err)
+	}
+
+	file, err := os.ReadFile(hc.filepath)
+	if err != nil {
+		return fmt.Errorf("error occurred while reading creds-file: %w", err)
+	}
+
+	// read passwords line by line to find the corresponding login
+	lines := strings.Split(string(file), "\n")
+	for i, line := range lines {
+		if tokens := strings.Split(line, ":"); len(tokens) == htpasswdValidTokensNumber {
+			if tokens[0] == login {
+				lines[i] = tokens[0] + ":" + string(newPassphrase)
+
+				break
+			}
+		}
+	}
+
+	// write new content to temporary file
+	// and replace the old file with temporary, so the operation is atomic
+	output := []byte(strings.Join(lines, "\n"))
+
+	tmpfile, err := os.CreateTemp(filepath.Dir(hc.filepath), "htpasswd-*.tmp")
+	if err != nil {
+		return fmt.Errorf("error occurred when creating temp htpasswd file: %w", err)
+	}
+
+	if _, err := tmpfile.Write(output); err != nil {
+		tmpfile.Close()
+		os.Remove(tmpfile.Name())
+		return fmt.Errorf("error occurred when writing to temp htpasswd file: %w", err)
+	}
+
+	if err := tmpfile.Close(); err != nil {
+		os.Remove(tmpfile.Name())
+		return fmt.Errorf("error occurred when closing temp htpasswd file: %w", err)
+	}
+
+	if err := os.Rename(tmpfile.Name(), hc.filepath); err != nil {
+		return fmt.Errorf("error occurred while replacing htpasswd file with new file: %w", err)
+	}
+
+	err = os.WriteFile(hc.filepath, output, constants.DefaultDirPerms)
```

Comment:
this should not be needed, right, or am I missing something?

you first write in tmpfile then you rename it to hc.filepath, that should be enough.